pkg/handlers/archive.go

```diff
@@ -233,39 +233,60 @@ func (a *Archive) ReadToMax(ctx context.Context, reader io.Reader) (data []byte,
 	return fileContent.Bytes(), nil
 }
 
+type mimeType string
+
 const (
-	arMimeType  = "application/x-unix-archive"
-	rpmMimeType = "application/x-rpm"
+	arMimeType  mimeType = "application/x-unix-archive"
+	rpmMimeType mimeType = "application/x-rpm"
 )
 
-// Define a map of mime types to corresponding command-line tools
-var mimeTools = map[string][]string{
+// mimeTools maps MIME types to the necessary command-line tools to handle them.
+// This centralizes the tool requirements for different file types.
+var mimeTools = map[mimeType][]string{
 	arMimeType:  {"ar"},
 	rpmMimeType: {"rpm2cpio", "cpio"},
 }
 
-// Check if the command-line tool is installed.
-func isToolInstalled(tool string) bool {
-	_, err := exec.LookPath(tool)
-	return err == nil
+// extractToolCache stores the availability of extraction tools, eliminating the need for repeated filesystem lookups.
+var extractToolCache map[string]bool
+
+func init() {
+	extractToolCache = make(map[string]bool, len(mimeTools))
+	// Preload the extractToolCache with the availability status of each required tool.
+	for _, tools := range mimeTools {
+		for _, tool := range tools {
+			extractToolCache[tool] = isToolInstalled(tool)
+		}
+	}
 }
 
-// Ensure all tools are available for given mime type.
-func ensureToolsForMimeType(mimeType string) error {
+func ensureToolsForMimeType(mimeType mimeType) error {
 	tools, exists := mimeTools[mimeType]
 	if !exists {
-		return fmt.Errorf("unsupported mime type")
+		return fmt.Errorf("unsupported mime type: %s", mimeType)
 	}
 
 	for _, tool := range tools {
 		if !isToolInstalled(tool) {
-			return fmt.Errorf("Required tool " + tool + " is not installed")
+			return fmt.Errorf("required tool %s is not installed", tool)
 		}
 	}
-
 	return nil
 }
 
+func isToolInstalled(tool string) bool {
+	if installed, ok := extractToolCache[tool]; ok {
+		return installed
+	}
+
+	// If the tool isn't in the cache, check its availability and cache it.
+	_, err := exec.LookPath(tool)
+	installed := err == nil
+	// Cache the result to avoid future filesystem checks for this tool.
+	extractToolCache[tool] = installed
```

Comment:
Writing to the map here makes this method not thread safe.

Reply:
But this is in the init method there is only a single thread? This is the only place writes occur, so I think this should be fine.

Reply:
or you mean during the run of the program?

Reply:
Yeah, during the run of the program. Is the method only going to be ran from `init()`? In that case I think we should inline it or add a comment so no one tries to re-use it as it's not obvious it could panic.

Reply:
> In that case I think we should inline it or add a comment so no one tries to re-use it as it's not obvious it could panic.

Call it [`secretIsToolInstalledDoNotUseOrYouWillBeFired`](https://github.com/facebook/react/blob/80bff5397bf854750dbe7c286f61654ea58938c5/src/umd/ReactUMDEntry.js). :P

Reply:
done